### ISSUE TITLE
Complex type using Ownsone as Primary key

### ISSUE THREAD

**Issue by windofny1** [dotnet/efcore#31426]
There is closed issue #23310  and there is example of code for using Owned type as Primary key. 
Know I have a problem with using Ownsone ValueObject in terms of DDD as Primary Key.
In my project there are two ValueObject types that can be used as keys:
OrdinaryValueObject (has key Value of Guid type)
TemporalValueObject (has key Value of Guid and SnapShopDate of DateTime).
Next I copy text from issue  #23310 and show implementations:

In my project there are classes for Entity, AggregateRoot and for ValueObject (in terms of DDD). And in Entity and Aggregate root there is property Id as ValueObject. And there are in my domain model exists some Entities with ValueObject as representation of Temporal row and it used as Primary Key. So here it is for example:
### Ordinary Value object (with one key =  Value as Guid)
```C#
public class OrdinaryValueObject : ValueObject, IOrdinaryValueObject
{
   public Guid Value { get; protected set; } 
  protected OrdinaryValueObject() : base()
    {

    }
    protected OrdinaryValueObject(Guid value)
    {
        Value = value;
    }
    protected OrdinaryValueObject(OrdinaryValueObject ordinaryValueObject)
    {
        Value = ordinaryValueObject.Value;
    }
    protected static OrdinaryValueObject CreateUniqueBase() 
    {
        return (new OrdinaryValueObject(Guid.NewGuid()));
    }
    public override IEnumerable<object> GetEqualityComponents()
    {
        yield return Value;
    }
}
```
### Temoral Value object (with two keys =  Value as Guid and SnapShotDate as DateTime)
```C#
public class TemporalValueObject : OrdinaryValueObject, ITemporalValueObject
    {
        public DateTime SnapShotDate { get; protected set; }

        protected TemporalValueObject():base() { }
        protected TemporalValueObject(Guid value, DateTime snapShotDate):base(value)
        { 
            SnapShotDate = snapShotDate;
        }
        protected TemporalValueObject(TemporalValueObject temporalValueObject):base(temporalValueObject.Value)
        { 
            SnapShotDate = temporalValueObject.SnapShotDate;
        }
        protected static TemporalValueObject CreateUniqueBase() 
        { 
            return new TemporalValueObject(Guid.NewGuid(), DateTime.UtcNow);
        }

        public override IEnumerable<object> GetEqualityComponents()
        {
            yield return Value;
            yield return SnapShotDate;
        }
    }
```
### Entity base class
```C#
public abstract class Entity<TId>:IEquatable<TId>, IEntity<TId>
        where TId : notnull 
    {
        
        public TId Id { get;  }
        protected Entity()
        {
           
        }
        protected Entity(TId id)
        {
            Id = id;
        }
        public override bool Equals(object? obj)
        {
            if (obj is null && Id is null)
            {
                return true;
            }            
            return obj is Entity<TId> entity && Id.Equals(entity.Id);
        }

        public bool Equals(TId? other)
        {
            return Equals((object?)other);
        }

        public static bool operator ==(Entity<TId> left, Entity<TId> right)
        {
            return Equals(left, right);
        }
        public static bool operator !=(Entity<TId> left, Entity<TId> right)
        {
            return !Equals(left, right);
        }
        public override int GetHashCode()
        {
            if (Id is null)
            {
                return 0;
            }
            return Id.GetHashCode();
        }
    }
```
### UserDataId temporal value object 
```C#
 public class UserDataId : TemporalValueObject
{
    private UserDataId() : base() { }
    private UserDataId(Guid value, DateTime snapShotDate) : base(value, snapShotDate)
    {

    }
    private UserDataId(TemporalValueObject temporalValueObject) : base(temporalValueObject) { }
    public static UserDataId CreateUnique()
    {
        return new(CreateUniqueBase());
    }

}
```
### UserData with temporal Id as UserId 
```C#
 
   public class UserData:Entity<UserDataId>
    {
         private UserData():base()
        {

        }
         private UserData(UserDataId userDataId, 
            string firstName,string middleName,string lastName, Email email, Phone phone,
            WorkingHoursEnum workingHours = WorkingHoursEnum.Normal, bool preventUpdate=false, DateTime? firedAt = null, DateTime? updatedDate=null,
            DivisionId? divisionId=null,PostId? postId=null, EmployeeNumber? employeeNumber=null, DateTime? dateOfBirth=null) :
            base(userDataId ?? UserDataId.CreateUnique())
        {
            PreventUpdate = preventUpdate;
            FiredAt = firedAt;
            FirstName = firstName;
            MiddleName = middleName;
            LastName = lastName;
            Email = email;
            Phone = phone;
            WorkingHours = workingHours;
            DivisionId = divisionId;
            PostId = postId;
            EmployeeNumber = employeeNumber;
            DateOfBirth = dateOfBirth;
            UpdatedDate = updatedDate;


        }
      
        public string FirstName { get;private set; }
        public string? MiddleName { get; private set; }
        public string LastName { get; private set; }

        public bool PreventUpdate { get; private set; }
        public DateTime? FiredAt { get; private set; }
        public DateTime? UpdatedDate { get; private set; }

        public Email Email { get; private set; }
        public Phone Phone { get; private set; }

      
        public DivisionId? DivisionId { get; private set; } = null;

        public PostId? PostId { get; private set; } = null;

        public EmployeeNumber? EmployeeNumber { get; private set; } = null;

        public DateTime? DateOfBirth { get; private set; } = null;
        public WorkingHoursEnum WorkingHours { get; set; } = WorkingHoursEnum.Normal;
     

        public static UserData Create(
            string firstName, string middleName, string lastName, Email email, Phone phone,
            WorkingHoursEnum workingHours=WorkingHoursEnum.Normal, bool preventUpdate=false, DateTime? firedAt=null, DateTime? updatedDate = null,
            DivisionId? divisionId = null, PostId? postId = null, EmployeeNumber? employeeNumber = null, 
            DateTime? dateOfBirth = null)
        {
            return new(UserDataId.CreateUnique(),  firstName, middleName, lastName,
                email, phone, workingHours, preventUpdate, firedAt, updatedDate, divisionId, postId, employeeNumber, dateOfBirth);
        }


    }

 
```
Also there is AggregateRoot (let's  omit description, we just need to know about is - it's contains User)
### User with entity - UserData
```C#
 public class User : AggregateRoot<UserId>
    {

        private User() : base(UserId.CreateUnique())
        {



        }

        public string UserName { get; private set; } = "";
        public string Password { get; private set; } = "";

        public DateTime CreatedDate { get; private set; }
        public DateTime? UpdatedDate { get; private set; }
        public DateTime? BlockedAt { get; private set; } = null;

        private readonly List<UserData> _userData = new();
        public IReadOnlyList<UserData> UserData => _userData.OrderByDescending(m => m.Id.SnapShotDate).ToList().AsReadOnly();
        public UserData? CurrentActualData => UserData.FirstOrDefault();

        private User(UserId userId, string userName, string password, List<UserData> userData, DateTime createdDate,
            DateTime? updatedDate = null, DateTime? blockedAt = null) : base(userId ?? UserId.CreateUnique())
        {
            UserName = userName;
            Password = password;
            BlockedAt = blockedAt;
            CreatedDate = createdDate;
            UpdatedDate = updatedDate;
            _userData = new List<UserData>(userData);
        }
        public static User Create(string userName, string password, List<UserData> userData, DateTime createdDate, DateTime? updatedDate = null, DateTime? blockedAt = null)
        {
            return new(UserId.CreateUnique(), userName, password, userData, createdDate, updatedDate, blockedAt);
        }
    }
```
Then was created extension methods for entity core configuration:

### Extensions 
```C#
  public static class ConfigurationsExtensions
    {
        public static string CreateForeignKey<T>(string Prefix="Id")
            where T:class
        {
            return $"{typeof(T).Name}{Prefix}";
        }
//When AggregateRoot contains Entity, that contains ValueObject as TemporalKey
        public static void OwnsOneAggregateToEntityKey<T, Y, TId, YId>(OwnedNavigationBuilder<T, Y> builder)
          where T : AggregateRoot<TId>
          where Y : Entity<YId>
          where TId : OrdinaryValueObject
          where YId : TemporalValueObject
        {
            
            builder.OwnsOne(x => x.Id, ownedBuilder => {
                ownedBuilder.Property(m => m.Value).HasColumnName("Id_Value");
                ownedBuilder.Property(m => m.SnapShotDate).HasColumnName("Id_SnapShotDate");
            });
            builder.Property(typeof(Guid), "Id_value").HasColumnName("Id_Value");
            builder.Property(typeof(DateTime), "Id_SnapShotDate").HasColumnName("Id_SnapShotDate");
            builder.HasKey("Id_value", "Id_SnapShotDate");
        }
//When AggregateRoot contains Entity, that contains ValueObject as default key (Guid value)
        public static void OwnsOneAggregateToValueObjectKey<T, TId>(EntityTypeBuilder<T> builder)
         where T : AggregateRoot<TId>
         where TId : OrdinaryValueObject

        {
            builder.OwnsOne(x => x.Id, ownedBuilder => {
                ownedBuilder.Property(x => x.Value).HasColumnName("Id_Value");
            });
            builder.Property(typeof(Guid), "Id_Value").HasColumnName("Id_Value");
            builder.HasKey("Id_Value");
        } 
    }
```
And then when use configuration:
```C#
  public class UserConfiguration : IEntityTypeConfiguration<User>
    {
        public void Configure(EntityTypeBuilder<User> builder)
        {
            ConfigureUserTable(builder);
            ConfigurationUserDataTable(builder);
        }

       
        private void ConfigurationUserDataTable(EntityTypeBuilder<User> builder)
        {
            builder.OwnsMany(m => m.UserData, ud =>
            {
              
                ud.ToTable("UserData"); 
        //extension methods used
                ud.WithOwner().HasForeignKey(ConfigurationsExtensions.CreateForeignKey<User>()); 
                ConfigurationsExtensions.OwnsOneAggregateToEntityKey<User,UserData, UserId, UserDataId>(ud);


                ud.Property(m => m.FirstName).HasMaxLength(255);
                ud.Property(m => m.MiddleName).HasMaxLength(255);
                ud.Property(m => m.LastName).HasMaxLength(255);

                ud.OwnsOne(m => m.Email);
                ud.OwnsOne(m => m.Phone);  
                ud.OwnsOne(m => m.EmployeeNumber);

            });
            builder.Metadata.FindNavigation(nameof(User.UserData))!.SetPropertyAccessMode(PropertyAccessMode.Field);
        }
        private void ConfigureUserTable(EntityTypeBuilder<User> builder)
        {
            builder.ToTable("Users"); 
 //extension methods used
            ConfigurationsExtensions.OwnsOneAggregateToValueObjectKey<User,UserId>(builder); 
            builder.Property(m => m.UserName).HasMaxLength(255);
            builder.Property(m => m.Password).HasMaxLength(50);
            builder.Ignore(m => m.CurrentActualData); 
        }
    }
```
 Whhen try to add to database some record of User - get error:
System.InvalidOperationException: Instances of entity types 'User' and 'UserId' are mapped to the same row with the key value '{Id_Value: 7BA746DE-4424-49FC-B7F9-34F87CA48381}', but have different property values '{Id_Value: 7BA746DE-4424-49FC-B7F9-34F87CA48381}' and '{Value: ce5a6d00-8aa7-4bf8-be3b-4bc31cbdded9}' for the column 'Id_Value'.
Thats means that property "Id_value" in User, that mapped for the ValueObject (OrdinaryValueObject) as Id.Value - not working. So for the UserData and UserDataId also not working.
What the right way to make complex object using in Ownsone as PrimaryKey ?

**Comment by windofny1** [comment]
@amantinband Need Your Help 🙏 )))

**Comment by ajcvickers** [comment]
This issue is lacking enough information for us to be able to fully understand what is happening. Please attach a small, runnable project or post a small, runnable code listing that reproduces what you are seeing so that we can investigate. The snippets above, when put together, still result in significant parts of missing code needed to make it run,

**Comment by windofny1** [comment]
> This issue is lacking enough information for us to be able to fully understand what is happening. Please attach a small, runnable project or post a small, runnable code listing that reproduces what you are seeing so that we can investigate. The snippets above, when put together, still result in significant parts of missing code needed to make it run,

@ajcvickers Thanks for Yours answer, I'll make separate project and leave here a link

**Comment by windofny1** [comment]
@ajcvickers Here it is:
https://github.com/windofny1/ExampleTemporaryKey

Please look at the:
[https://github.com/windofny1/ExampleTemporaryKey/blob/main/src/infrastructure/TemporalKey.Infastracture/Persistence/Configurations/UserConfiguration.cs](url)
In 
`
 private void ConfigurationUserDataTable(EntityTypeBuilder<User> builder)
`
This part of code:
`
ud.OwnsOne(x => x.Id, ownedBuilder => {
                    ownedBuilder.Property(m => m.Value).HasColumnName("Id_Value");
                    ownedBuilder.Property(m => m.SnapShotDate).HasColumnName("Id_SnapShotDate");
                });
                builder.Property(typeof(Guid), "Id_value").HasColumnName("Id_Value");
                builder.Property(typeof(DateTime), "Id_SnapShotDate").HasColumnName("Id_SnapShotDate");
                builder.HasKey("Id_value", "Id_SnapShotDate");
`
That what I'am trying to make - UserData must be with key of two columns - Guid and DateTime in the database table. But at the code part it must be ValueObject with name Id.
And also need connect with Parent Entity - User.
If now try to create migration:
`add-migration InitialCreate -StartupProject TemporalKey.WebAPI`
And get error:
Unable to create a 'DbContext' of type ''. The exception 'The relationship from 'UserData' to 'User.UserData' with foreign key properties {'UserId' : Guid} cannot target the primary key {'Id_value' : Guid, 'Id_SnapShotDate' : DateTime} because it is not compatible. Configure a principal key or a set of foreign key properties with compatible types for this relationship.' was thrown while attempting to create an instance. For the different patterns supported at design time, see https://go.microsoft.com/fwlink/?linkid=851728
 
 In other projects I use PrimaryKey as ValueObject with one property inside, use HasConvertion and all works like a charm. But what if I need PrimaryKey as ValueObject with two or more properties ? That what problem is. 
 P.s.: I know that Entity core can work with Temporal Tables, but want to know may I control Primary Key for Temporal tables by my self and use custom primary keys with two columns.
 And Thank You for Your time and attention to problem!

**Comment by ajcvickers** [comment]
@windofny1 `UserId` is both mapped as a non-entity type through its use as a key, and then also as an entity type in the call to `OwnsOne`. This is not supported by EF Core. A type used as a property type cannot also be used as an entity type.

Mapping of collections of FKs is tracked by #31845.

**Comment by windofny1** [comment]
> @windofny1 `UserId` is both mapped as a non-entity type through its use as a key, and then also as an entity type in the call to `OwnsOne`. This is not supported by EF Core. A type used as a property type cannot also be used as an entity type.
> 
> Mapping of collections of FKs is tracked by #31845.

Thanks for Your time! Will use another approarch.

**Comment by savaladaojr** [comment]
> There is closed issue #23310 and there is example of code for using Owned type as Primary key. Know I have a problem with using Ownsone ValueObject in terms of DDD as Primary Key. In my project there are two ValueObject types that can be used as keys: OrdinaryValueObject (has key Value of Guid type) TemporalValueObject (has key Value of Guid and SnapShopDate of DateTime). Next I copy text from issue #23310 and show implementations:
> 
> In my project there are classes for Entity, AggregateRoot and for ValueObject (in terms of DDD). And in Entity and Aggregate root there is property Id as ValueObject. And there are in my domain model exists some Entities with ValueObject as representation of Temporal row and it used as Primary Key. So here it is for example:
> 
> ### Ordinary Value object (with one key = Value as Guid)
> ```cs
> public class OrdinaryValueObject : ValueObject, IOrdinaryValueObject
> {
>    public Guid Value { get; protected set; } 
>   protected OrdinaryValueObject() : base()
>     {
> 
>     }
>     protected OrdinaryValueObject(Guid value)
>     {
>         Value = value;
>     }
>     protected OrdinaryValueObject(OrdinaryValueObject ordinaryValueObject)
>     {
>         Value = ordinaryValueObject.Value;
>     }
>     protected static OrdinaryValueObject CreateUniqueBase() 
>     {
>         return (new OrdinaryValueObject(Guid.NewGuid()));
>     }
>     public override IEnumerable<object> GetEqualityComponents()
>     {
>         yield return Value;
>     }
> }
> ```
> 
> ### Temoral Value object (with two keys = Value as Guid and SnapShotDate as DateTime)
> ```cs
> public class TemporalValueObject : OrdinaryValueObject, ITemporalValueObject
>     {
>         public DateTime SnapShotDate { get; protected set; }
> 
>         protected TemporalValueObject():base() { }
>         protected TemporalValueObject(Guid value, DateTime snapShotDate):base(value)
>         { 
>             SnapShotDate = snapShotDate;
>         }
>         protected TemporalValueObject(TemporalValueObject temporalValueObject):base(temporalValueObject.Value)
>         { 
>             SnapShotDate = temporalValueObject.SnapShotDate;
>         }
>         protected static TemporalValueObject CreateUniqueBase() 
>         { 
>             return new TemporalValueObject(Guid.NewGuid(), DateTime.UtcNow);
>         }
> 
>         public override IEnumerable<object> GetEqualityComponents()
>         {
>             yield return Value;
>             yield return SnapShotDate;
>         }
>     }
> ```
> 
> ### Entity base class
> ```cs
> public abstract class Entity<TId>:IEquatable<TId>, IEntity<TId>
>         where TId : notnull 
>     {
>         
>         public TId Id { get;  }
>         protected Entity()
>         {
>            
>         }
>         protected Entity(TId id)
>         {
>             Id = id;
>         }
>         public override bool Equals(object? obj)
>         {
>             if (obj is null && Id is null)
>             {
>                 return true;
>             }            
>             return obj is Entity<TId> entity && Id.Equals(entity.Id);
>         }
> 
>         public bool Equals(TId? other)
>         {
>             return Equals((object?)other);
>         }
> 
>         public static bool operator ==(Entity<TId> left, Entity<TId> right)
>         {
>             return Equals(left, right);
>         }
>         public static bool operator !=(Entity<TId> left, Entity<TId> right)
>         {
>             return !Equals(left, right);
>         }
>         public override int GetHashCode()
>         {
>             if (Id is null)
>             {
>                 return 0;
>             }
>             return Id.GetHashCode();
>         }
>     }
> ```
> 
> ### UserDataId temporal value object
> ```cs
>  public class UserDataId : TemporalValueObject
> {
>     private UserDataId() : base() { }
>     private UserDataId(Guid value, DateTime snapShotDate) : base(value, snapShotDate)
>     {
> 
>     }
>     private UserDataId(TemporalValueObject temporalValueObject) : base(temporalValueObject) { }
>     public static UserDataId CreateUnique()
>     {
>         return new(CreateUniqueBase());
>     }
> 
> }
> ```
> 
> ### UserData with temporal Id as UserId
> ```cs
>  
>    public class UserData:Entity<UserDataId>
>     {
>          private UserData():base()
>         {
> 
>         }
>          private UserData(UserDataId userDataId, 
>             string firstName,string middleName,string lastName, Email email, Phone phone,
>             WorkingHoursEnum workingHours = WorkingHoursEnum.Normal, bool preventUpdate=false, DateTime? firedAt = null, DateTime? updatedDate=null,
>             DivisionId? divisionId=null,PostId? postId=null, EmployeeNumber? employeeNumber=null, DateTime? dateOfBirth=null) :
>             base(userDataId ?? UserDataId.CreateUnique())
>         {
>             PreventUpdate = preventUpdate;
>             FiredAt = firedAt;
>             FirstName = firstName;
>             MiddleName = middleName;
>             LastName = lastName;
>             Email = email;
>             Phone = phone;
>             WorkingHours = workingHours;
>             DivisionId = divisionId;
>             PostId = postId;
>             EmployeeNumber = employeeNumber;
>             DateOfBirth = dateOfBirth;
>             UpdatedDate = updatedDate;
> 
> 
>         }
>       
>         public string FirstName { get;private set; }
>         public string? MiddleName { get; private set; }
>         public string LastName { get; private set; }
> 
>         public bool PreventUpdate { get; private set; }
>         public DateTime? FiredAt { get; private set; }
>         public DateTime? UpdatedDate { get; private set; }
> 
>         public Email Email { get; private set; }
>         public Phone Phone { get; private set; }
> 
>       
>         public DivisionId? DivisionId { get; private set; } = null;
> 
>         public PostId? PostId { get; private set; } = null;
> 
>         public EmployeeNumber? EmployeeNumber { get; private set; } = null;
> 
>         public DateTime? DateOfBirth { get; private set; } = null;
>         public WorkingHoursEnum WorkingHours { get; set; } = WorkingHoursEnum.Normal;
>      
> 
>         public static UserData Create(
>             string firstName, string middleName, string lastName, Email email, Phone phone,
>             WorkingHoursEnum workingHours=WorkingHoursEnum.Normal, bool preventUpdate=false, DateTime? firedAt=null, DateTime? updatedDate = null,
>             DivisionId? divisionId = null, PostId? postId = null, EmployeeNumber? employeeNumber = null, 
>             DateTime? dateOfBirth = null)
>         {
>             return new(UserDataId.CreateUnique(),  firstName, middleName, lastName,
>                 email, phone, workingHours, preventUpdate, firedAt, updatedDate, divisionId, postId, employeeNumber, dateOfBirth);
>         }
> 
> 
>     }
> 
>  
> ```
> 
> Also there is AggregateRoot (let's omit description, we just need to know about is - it's contains User)
> 
> ### User with entity - UserData
> ```cs
>  public class User : AggregateRoot<UserId>
>     {
> 
>         private User() : base(UserId.CreateUnique())
>         {
> 
> 
> 
>         }
> 
>         public string UserName { get; private set; } = "";
>         public string Password { get; private set; } = "";
> 
>         public DateTime CreatedDate { get; private set; }
>         public DateTime? UpdatedDate { get; private set; }
>         public DateTime? BlockedAt { get; private set; } = null;
> 
>         private readonly List<UserData> _userData = new();
>         public IReadOnlyList<UserData> UserData => _userData.OrderByDescending(m => m.Id.SnapShotDate).ToList().AsReadOnly();
>         public UserData? CurrentActualData => UserData.FirstOrDefault();
> 
>         private User(UserId userId, string userName, string password, List<UserData> userData, DateTime createdDate,
>             DateTime? updatedDate = null, DateTime? blockedAt = null) : base(userId ?? UserId.CreateUnique())
>         {
>             UserName = userName;
>             Password = password;
>             BlockedAt = blockedAt;
>             CreatedDate = createdDate;
>             UpdatedDate = updatedDate;
>             _userData = new List<UserData>(userData);
>         }
>         public static User Create(string userName, string password, List<UserData> userData, DateTime createdDate, DateTime? updatedDate = null, DateTime? blockedAt = null)
>         {
>             return new(UserId.CreateUnique(), userName, password, userData, createdDate, updatedDate, blockedAt);
>         }
>     }
> ```
> 
> Then was created extension methods for entity core configuration:
> 
> ### Extensions
> ```cs
>   public static class ConfigurationsExtensions
>     {
>         public static string CreateForeignKey<T>(string Prefix="Id")
>             where T:class
>         {
>             return $"{typeof(T).Name}{Prefix}";
>         }
> //When AggregateRoot contains Entity, that contains ValueObject as TemporalKey
>         public static void OwnsOneAggregateToEntityKey<T, Y, TId, YId>(OwnedNavigationBuilder<T, Y> builder)
>           where T : AggregateRoot<TId>
>           where Y : Entity<YId>
>           where TId : OrdinaryValueObject
>           where YId : TemporalValueObject
>         {
>             
>             builder.OwnsOne(x => x.Id, ownedBuilder => {
>                 ownedBuilder.Property(m => m.Value).HasColumnName("Id_Value");
>                 ownedBuilder.Property(m => m.SnapShotDate).HasColumnName("Id_SnapShotDate");
>             });
>             builder.Property(typeof(Guid), "Id_value").HasColumnName("Id_Value");
>             builder.Property(typeof(DateTime), "Id_SnapShotDate").HasColumnName("Id_SnapShotDate");
>             builder.HasKey("Id_value", "Id_SnapShotDate");
>         }
> //When AggregateRoot contains Entity, that contains ValueObject as default key (Guid value)
>         public static void OwnsOneAggregateToValueObjectKey<T, TId>(EntityTypeBuilder<T> builder)
>          where T : AggregateRoot<TId>
>          where TId : OrdinaryValueObject
> 
>         {
>             builder.OwnsOne(x => x.Id, ownedBuilder => {
>                 ownedBuilder.Property(x => x.Value).HasColumnName("Id_Value");
>             });
>             builder.Property(typeof(Guid), "Id_Value").HasColumnName("Id_Value");
>             builder.HasKey("Id_Value");
>         } 
>     }
> ```
> 
> And then when use configuration:
> 
> ```cs
>   public class UserConfiguration : IEntityTypeConfiguration<User>
>     {
>         public void Configure(EntityTypeBuilder<User> builder)
>         {
>             ConfigureUserTable(builder);
>             ConfigurationUserDataTable(builder);
>         }
> 
>        
>         private void ConfigurationUserDataTable(EntityTypeBuilder<User> builder)
>         {
>             builder.OwnsMany(m => m.UserData, ud =>
>             {
>               
>                 ud.ToTable("UserData"); 
>         //extension methods used
>                 ud.WithOwner().HasForeignKey(ConfigurationsExtensions.CreateForeignKey<User>()); 
>                 ConfigurationsExtensions.OwnsOneAggregateToEntityKey<User,UserData, UserId, UserDataId>(ud);
> 
> 
>                 ud.Property(m => m.FirstName).HasMaxLength(255);
>                 ud.Property(m => m.MiddleName).HasMaxLength(255);
>                 ud.Property(m => m.LastName).HasMaxLength(255);
> 
>                 ud.OwnsOne(m => m.Email);
>                 ud.OwnsOne(m => m.Phone);  
>                 ud.OwnsOne(m => m.EmployeeNumber);
> 
>             });
>             builder.Metadata.FindNavigation(nameof(User.UserData))!.SetPropertyAccessMode(PropertyAccessMode.Field);
>         }
>         private void ConfigureUserTable(EntityTypeBuilder<User> builder)
>         {
>             builder.ToTable("Users"); 
>  //extension methods used
>             ConfigurationsExtensions.OwnsOneAggregateToValueObjectKey<User,UserId>(builder); 
>             builder.Property(m => m.UserName).HasMaxLength(255);
>             builder.Property(m => m.Password).HasMaxLength(50);
>             builder.Ignore(m => m.CurrentActualData); 
>         }
>     }
> ```
> 
> Whhen try to add to database some record of User - get error: System.InvalidOperationException: Instances of entity types 'User' and 'UserId' are mapped to the same row with the key value '{Id_Value: 7BA746DE-4424-49FC-B7F9-34F87CA48381}', but have different property values '{Id_Value: 7BA746DE-4424-49FC-B7F9-34F87CA48381}' and '{Value: ce5a6d00-8aa7-4bf8-be3b-4bc31cbdded9}' for the column 'Id_Value'. Thats means that property "Id_value" in User, that mapped for the ValueObject (OrdinaryValueObject) as Id.Value - not working. So for the UserData and UserDataId also not working. What the right way to make complex object using in Ownsone as PrimaryKey ?

awesome, thanks you so much.
